validation/src/lib.rs

```diff
@@ -383,25 +412,52 @@ impl<C, N, P> ParachainValidation<C, N, P> where
 				validation_para,
 				relay_parent,
 				collators,
-				client,
+				client.clone(),
 				max_block_data_size,
 			);
 
 			collation_work.then(move |result| match result {
-				Ok((collation, extrinsic)) => {
+				Ok((mut collation, extrinsic)) => {
+					let chunks = erasure::obtain_chunks(authorities_num,
+														&collation.pov.block_data,
+														&extrinsic.clone()).unwrap();
+
+					let chunks_ref: Vec<_> = chunks.iter().map(|c| &c[..]).collect();
+					let branches = erasure::branches(chunks_ref.clone());
+					let root = branches.root();
+					let proofs: Vec<_> = branches.map(|(proof, _)| proof).collect();
+					collation.receipt.erasure_root = Some(root);
+
+					let candidate_hash = collation.receipt.hash();
+
+					// As a node that has received a valid collation, we store all of the
+					// erasure coded chunks pieces, at least for now.
+					let chunks: Vec<_> = chunks.into_iter().zip(proofs)
+						.enumerate()
+						.map(|(index, (chunk, proof))| ErasureChunk {
+							relay_parent,
+							candidate_hash,
+							chunk,
+							index: index as u32,
```

Comment:
suggest changing the argument types here to type aliases instead, as mentioned here https://github.com/paritytech/polkadot/pull/345/files#r308998575